src/d/actor/d_a_obj_hat.cpp

```diff
@@ -4,91 +4,238 @@
+        /* SrcGObjAt Mtrl    */ 0,
+        /* SrcGObjAt SPrm    */ 0,//AT_TYPE_WIND,
+        /* SrcGObjTg Se      */ 0,
+        /* SrcGObjTg HitMark */ 0,
+        /* SrcGObjTg Spl     */ 0,
+        /* SrcGObjTg Mtrl    */ 0,
+        /* SrcGObjTg SPrm    */ dCcG_TgSPrm_NoHitMark_e | dCcG_TgSPrm_NoConHit_e,
+        /* SrcGObjCo SPrm    */ 0,
+    },
+    // cM3dGCylS
+    {
+        /* Center */ 0.0f,
+        0.0f,
+        0.0f,
+        /* Radius */ 70.0f,
+        /* Height */ 80.0f,
+    },
+};
+
 /* 0000045C-000005D4       .text createHeap__10daObjHat_cFv */
-void daObjHat_c::createHeap() {
-    /* Nonmatching */
+BOOL daObjHat_c::createHeap() {
+    J3DModelData* pModelData = (J3DModelData*)dComIfG_getObjectIDRes("Ro", l_bmd_ix_tbl[this->id]);
+    if (pModelData == NULL) {
+        return FALSE;
+    } else {
+        mDoExt_McaMorf* morf = new mDoExt_McaMorf(
+            pModelData, NULL, NULL,
+            (J3DAnmTransformKey*)dComIfG_getObjectIDRes("Ro", l_bck_ix_tbl[this->id]),
+            J3DFrameCtrl::LOOP_REPEAT_e, 1.0f, 0, -1, 1, NULL, 0x80000, 0x37441422);
+
+        this->mpMorf = morf;
+        if (this->mpMorf == NULL || this->mpMorf->getModel() == NULL) {
+            return FALSE;
+        } else {
+            this->model = this->mpMorf->getModel();
+            acch_cir.SetWall(30.0f, 30.0f);
+            acch.Set(&this->current.pos, &this->old.pos, this, 1, &acch_cir, &this->speed,
+                     &this->current.angle, &this->shape_angle);
+            return TRUE;
+        }
+    }
+    return FALSE;
 }
 
 /* 000005D4-000006AC       .text createInit__10daObjHat_cFv */
-void daObjHat_c::createInit() {
-    /* Nonmatching */
+s32 daObjHat_c::createInit() {
+    stts.Init(2, 0xff, this);
+    cyl.Set(l_cyl_src);
+    this->cyl.SetStts(&this->stts);
+    this->cullMtx = this->model->mBaseTransformMtx;
+    fopAcM_setCullSizeBox(this, -50.0f, 0.0f, -50.0f, 50.0f, 200.0f, 50.0f);
+    this->gravity = -5.0f;
+
+    fopAc_ac_c* parent;
+    int res = fopAcM_SearchByID(this->parentActorID, &parent);
+    if ((res != 0) && (parent != NULL)) {
+        setSpeed(static_cast<daNpcRoten_c*>(parent)->getWindVec());
+    }
+    setMtx();
+    return cPhs_COMPLEATE_e;
 }
 
 /* 000006AC-000006DC       .text _delete__10daObjHat_cFv */
 bool daObjHat_c::_delete() {
-    /* Nonmatching */
+    dComIfG_resDelete(&mPhs, "Ro");
+    return TRUE;
 }
 
 /* 000006DC-0000073C       .text _draw__10daObjHat_cFv */
 bool daObjHat_c::_draw() {
-    /* Nonmatching */
+    g_env_light.settingTevStruct(TEV_TYPE_ACTOR, &this->current.pos, &this->tevStr);
+    g_env_light.setLightTevColorType(this->model, &this->tevStr);
+    this->mpMorf->updateDL();
+    return TRUE;
 }
 
+static daObjHat_c::MoveFunc_t moveProc[] = {
+    &daObjHat_c::executeNormal,
+};
+
+
 /* 0000073C-00000884       .text _execute__10daObjHat_cFv */
 bool daObjHat_c::_execute() {
-    /* Nonmatching */
+    // dCcD_Sph::~dCcD_Sp
+
+    (this->*moveProc[this->procid])();
```

Comment:
"Proc ID" means something different in this codebase (see `fpc_ProcID`). This should be called the state or similar.
```suggestion
    (this->*moveProc[this->mState])();
```